src/meta/meta_ckpt.c

```diff
@@ -1270,31 +1416,69 @@ int
 __wt_meta_ckptlist_set(
   WT_SESSION_IMPL *session, WT_DATA_HANDLE *dhandle, WT_CKPT *ckptbase, WT_LSN *ckptlsn)
 {
+    WT_BLOCK_MODS *blk;
     WT_CKPT *ckpt;
     WT_DECL_ITEM(buf);
     WT_DECL_RET;
-    const char *fname;
-    bool has_lsn;
+    WT_ITEM file_blkmods_buffer, checkpoint_blkmods_buffer;
+    u_int i;
+    const char *filename;
+    bool blkmods_are_ok, has_lsn;
 
-    fname = dhandle->name;
-    WT_RET(__wt_scr_alloc(session, 1024, &buf));
+    filename = dhandle->name;
+    blkmods_are_ok = true;
+
+    WT_ERR(__wt_scr_alloc(session, 1024, &buf));
     WT_ERR(__wt_meta_ckptlist_to_meta(session, ckptbase, buf));
     /* Add backup block modifications for any added checkpoint. */
     WT_CKPT_FOREACH (ckptbase, ckpt)
-        if (F_ISSET(ckpt, WT_CKPT_ADD))
+        if (F_ISSET(ckpt, WT_CKPT_ADD)) {
+            WT_CLEAR(file_blkmods_buffer);
+            WT_CLEAR(checkpoint_blkmods_buffer);
+
             WT_ERR(__wt_ckpt_blkmod_to_meta(session, buf, ckpt));
 
+            if (strcmp(filename, WT_HS_URI) == 0 || session->hs_cursor_counter != 0) {
+                /* Skip history store tables */
+            } else if (F_ISSET(dhandle, WT_DHANDLE_IS_METADATA)) {
+                /* Skip */
+            } else {
+                for (i = 0, blk = &ckpt->backup_blocks[0]; i < WT_BLKINCR_MAX; ++i, ++blk) {
+                    if (!F_ISSET(blk, WT_BLOCK_MODS_VALID))
+                        continue;
+
+                    WT_ERR(
+                      ckpt_blkmod_to_item(session, ckpt, blk->id_str, &checkpoint_blkmods_buffer));
+
+                    ret = get_blkmods(session, filename, blk->id_str, &file_blkmods_buffer);
+                    if (file_blkmods_buffer.size > 0 && checkpoint_blkmods_buffer.size > 0) {
+                        blkmods_are_ok = false;
+                        ret = check_incorrect_modified_bits(
+                          &file_blkmods_buffer, &checkpoint_blkmods_buffer, &blkmods_are_ok);
+
+                        if ((ret != 0) || !blkmods_are_ok)
+                            WT_ERR_MSG(session, EINVAL,
+                              "File blkmods are not compatible with those in the checkpoint");
```

Comment:
I think the other function should assert and panic and drop core at the moment the problem is detected. I don't think it should be an error return.

Also should this be under HAVE_DIAGNOSTIC? I don't remember what was discussed originally but I'd be concerned that for a lot of tables on a release build it may add checkpoint processing time.

Reply:
I am fine with it not being DIAGNOSTIC unless we see perf impact.